Source/lighting.h

```diff
@@ -74,10 +75,113 @@ void ChangeVisionXY(int id, Point position);
 void ProcessVisionList();
 void lighting_color_cycling();
 
+template <typename F>
+auto CrawlFlipsX(Displacement mirrored, F function) -> invoke_result_t<decltype(function), Displacement>
+{
+	const Displacement Flips[] = { mirrored.flipX(), mirrored };
+	for (auto displacement : Flips) {
+		auto ret = function(displacement);
+		if (ret)
+			return ret;
+	}
+	return {};
+}
+
+template <typename F>
+auto CrawlFlipsY(Displacement mirrored, F function) -> invoke_result_t<decltype(function), Displacement>
+{
+	const Displacement Flips[] = { mirrored, mirrored.flipY() };
+	for (auto displacement : Flips) {
+		auto ret = function(displacement);
+		if (ret)
+			return ret;
+	}
+	return {};
+}
+
+template <typename F>
+auto CrawlFlipsXY(Displacement mirrored, F function) -> invoke_result_t<decltype(function), Displacement>
+{
+	const Displacement Flips[] = { mirrored.flipX(), mirrored, mirrored.flipXY(), mirrored.flipY() };
+	for (auto displacement : Flips) {
+		auto ret = function(displacement);
+		if (ret)
+			return ret;
+	}
+	return {};
+}
+
+constexpr int MaxCrawlRadius = 18;
+
+/**
+ * CrawlTable specifies X- and Y-coordinate deltas from a missile target coordinate.
+ *
+ * n=4
+ *
+ *    y
+ *    ^
+ *    |  1
+ *    | 3#4
+ *    |  2
+ *    +-----> x
+ *
+ * n=16
+ *
+ *    y
+ *    ^
+ *    |  314
+ *    | B7 8C
+ *    | F # G
+ *    | D9 AE
+ *    |  526
+ *    +-------> x
+ */
+
+template <typename F>

```

Comment:
Perhaps we can avoid templates by making it accept a `FunctionRef<bool(Displacement)>` where bool indicates whether to abort or not. This will also result in less generated code. The return value stuff can be managed by the caller.

We'd need to add a `FunctionRef` class first (e.g. from https://github.com/TartanLlama/function_ref)

Reply:
We can, but this lends itself to a nastier API, i. e.:

Reply:
We could have only a single templated function in the header, with the rest in the cpp file, e.g.:

```c++
void Crawl(unsigned minRadius, unsigned maxRadius, tl::function_ref<bool(Displacement)>);

template <typename T>
std::optional<T> Crawl(unsigned minRadius, unsigned maxRadius, tl::function_ref<std::optional<T>(Displacement)> fn) {
	std::optional<T> result
    Crawl(minRadius, maxRadius, [&result, &fn](Displacement displacement) {
      result = fn(displacement);
      return !result;
    });
  return result;
}

Reply:
I guess it's OK, but definitely not as part of this PR.

Reply:
I'll send a PR for this shortly